File: frontend/runtime_console_v3/src/lib/seller-rbac.ts
export type PermissionMatrix = Record<string, Record<string, boolean>>;

export const ROLE_LABELS: Record<string, string> = {
  store_owner: "Proprietário",
  manager: "Gerente",
  operator: "Operador",
  stock_keeper: "Estoquista",
  support: "Suporte",
  finance: "Financeiro",
  marketing: "Marketing",
};

const ALL_ACTIONS = ["view", "create", "edit", "delete", "export", "approve"] as const;
const MODULES = [
  "orders",
  "customers",
  "inventory",
  "catalog",
  "tickets",
  "finance",
  "settings",
  "team",
  "marketing",
] as const;

function fullAccess(): PermissionMatrix {
  return Object.fromEntries(
    MODULES.map((m) => [m, Object.fromEntries(ALL_ACTIONS.map((a) => [a, true]))]),
  );
}

export const ROLE_DEFAULTS: Record<string, PermissionMatrix> = {
  store_owner: fullAccess(),
  manager: {
    orders: { view: true, create: true, edit: true, delete: false, export: true, approve: true },
    customers: { view: true, create: true, edit: true, delete: false, export: true, approve: false },
    inventory: { view: true, create: true, edit: true, delete: false, export: false, approve: false },
    catalog: { view: true, create: true, edit: true, delete: false, export: false, approve: false },
    tickets: { view: true, create: true, edit: true, delete: false, export: false, approve: false },
    finance: { view: true, create: false, edit: false, delete: false, export: true, approve: false },
    settings: { view: true, create: false, edit: false, delete: false, export: false, approve: false },
    team: { view: true, create: false, edit: false, delete: false, export: false, approve: false },
    marketing: { view: true, create: true, edit: true, delete: false, export: true, approve: false },
  },
  operator: {
    orders: { view: true, create: false, edit: true, delete: false, export: false, approve: false },
    customers: { view: true, create: false, edit: false, delete: false, export: false, approve: false },
    inventory: { view: true, create: false, edit: true, delete: false, export: false, approve: false },
    catalog: { view: true, create: false, edit: false, delete: false, export: false, approve: false },
    tickets: { view: true, create: false, edit: false, delete: false, export: false, approve: false },
    finance: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    settings: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    team: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    marketing: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
  },
  stock_keeper: {
    orders: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    customers: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    inventory: { view: true, create: true, edit: true, delete: false, export: false, approve: false },
    catalog: { view: true, create: false, edit: false, delete: false, export: false, approve: false },
    tickets: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    finance: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    settings: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    team: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    marketing: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
  },
  support: {
    orders: { view: true, create: false, edit: false, delete: false, export: false, approve: false },
    customers: { view: true, create: false, edit: false, delete: false, export: false, approve: false },
    inventory: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    catalog: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    tickets: { view: true, create: true, edit: true, delete: false, export: false, approve: false },
    finance: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    settings: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    team: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    marketing: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
  },
  finance: {
    orders: { view: true, create: false, edit: false, delete: false, export: true, approve: false },
    customers: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    inventory: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    catalog: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    tickets: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    finance: { view: true, create: false, edit: false, delete: false, export: true, approve: false },
    settings: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    team: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    marketing: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
  },
  marketing: {
    orders: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    customers: { view: true, create: false, edit: false, delete: false, export: true, approve: false },
    inventory: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    catalog: { view: true, create: false, edit: false, delete: false, export: false, approve: false },
    tickets: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    finance: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    settings: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    team: { view: false, create: false, edit: false, delete: false, export: false, approve: false },
    marketing: { view: true, create: true, edit: true, delete: false, export: true, approve: false },
  },
};

export const MODULE_LABELS: Record<string, string> = {
  orders: "Pedidos",
  customers: "Clientes",
  inventory: "Estoque",
  catalog: "Catálogo",
  tickets: "Tickets",
  finance: "Financeiro",
  settings: "Configurações",
  team: "Equipe",
  marketing: "Marketing",
};

export const ACTION_LABELS: Record<string, string> = {
  view: "Visualizar",
  create: "Criar",
  edit: "Editar",
  delete: "Excluir",
  export: "Exportar",
  approve: "Aprovar",
};

export function mergePermissions(
  role: string,
  override?: PermissionMatrix | null,
): PermissionMatrix {
  const base = structuredClone(ROLE_DEFAULTS[role] ?? ROLE_DEFAULTS.operator);
  if (!override) return base;
  for (const [module, actions] of Object.entries(override)) {
    if (!base[module]) base[module] = {};
    for (const [action, allowed] of Object.entries(actions)) {
      base[module][action] = Boolean(allowed);
    }
  }
  return base;
}

export function hasPermission(
  role: string,
  permissions: PermissionMatrix | null | undefined,
  module: string,
  action: string,
): boolean {
  if (role === "store_owner") return true;
  const matrix = permissions ?? mergePermissions(role, null);
  return Boolean(matrix[module]?.[action]);
}

File: services/api/app/marketplace/seller_rbac.py
"""RBAC granular do painel lojista (equipe da loja)."""

from __future__ import annotations

from copy import deepcopy
from typing import Any

# Módulos × ações
PermissionMatrix = dict[str, dict[str, bool]]

_ACTIONS = ("view", "create", "edit", "delete", "export", "approve")
_DENY = dict.fromkeys(_ACTIONS, False)

ROLE_DEFAULTS: dict[str, PermissionMatrix] = {
    "store_owner": {
        m: {a: True for a in ("view", "create", "edit", "delete", "export", "approve")}
        for m in ("orders", "customers", "inventory", "catalog", "tickets", "finance", "settings", "team", "marketing")
    },
    "manager": {
        "orders": {"view": True, "create": True, "edit": True, "delete": False, "export": True, "approve": True},
        "customers": {"view": True, "create": True, "edit": True, "delete": False, "export": True, "approve": False},
        "inventory": {"view": True, "create": True, "edit": True, "delete": False, "export": False, "approve": False},
        "catalog": {"view": True, "create": True, "edit": True, "delete": False, "export": False, "approve": False},
        "tickets": {"view": True, "create": True, "edit": True, "delete": False, "export": False, "approve": False},
        "finance": {"view": True, "create": False, "edit": False, "delete": False, "export": True, "approve": False},
        "settings": {"view": True, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "team": {"view": True, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "marketing": {"view": True, "create": True, "edit": True, "delete": False, "export": True, "approve": False},
    },
    "operator": {
        "orders": {"view": True, "create": False, "edit": True, "delete": False, "export": False, "approve": False},
        "customers": {"view": True, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "inventory": {"view": True, "create": False, "edit": True, "delete": False, "export": False, "approve": False},
        "catalog": {"view": True, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "tickets": {"view": True, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "finance": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "settings": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "team": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "marketing": _DENY,
    },
    "stock_keeper": {
        "orders": _DENY,
        "customers": _DENY,
        "inventory": {"view": True, "create": True, "edit": True, "delete": False, "export": False, "approve": False},
        "catalog": {"view": True, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "tickets": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "finance": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "settings": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "team": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "marketing": _DENY,
    },
    "support": {
        "orders": {"view": True, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "customers": {"view": True, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "inventory": _DENY,
        "catalog": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "tickets": {"view": True, "create": True, "edit": True, "delete": False, "export": False, "approve": False},
        "finance": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "settings": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "team": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "marketing": _DENY,
    },
    "finance": {
        "orders": {"view": True, "create": False, "edit": False, "delete": False, "export": True, "approve": False},
        "customers": _DENY,
        "inventory": _DENY,
        "catalog": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "tickets": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "finance": {"view": True, "create": False, "edit": False, "delete": False, "export": True, "approve": False},
        "settings": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "team": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "marketing": _DENY,
    },
    "marketing": {
        "orders": _DENY,
        "customers": {"view": True, "create": False, "edit": False, "delete": False, "export": True, "approve": False},
        "inventory": _DENY,
        "catalog": {"view": True, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "tickets": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "finance": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "settings": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "team": {"view": False, "create": False, "edit": False, "delete": False, "export": False, "approve": False},
        "marketing": {"view": True, "create": True, "edit": True, "delete": False, "export": True, "approve": False},
    },
}

VALID_ROLES = frozenset(ROLE_DEFAULTS.keys())


def merge_permissions(role: str, override: dict[str, Any] | None) -> PermissionMatrix:
    base = deepcopy(ROLE_DEFAULTS.get(role, ROLE_DEFAULTS["operator"]))
    if not override:
        return base
    for module, actions in override.items():
        if module not in base:
            base[module] = {}
        if isinstance(actions, dict):
            for action, allowed in actions.items():
                base[module][action] = bool(allowed)
    return base


def has_store_permission(
    role: str,
    permissions: dict[str, Any] | None,
    module: str,
    action: str,
) -> bool:
    if role == "store_owner":
        return True
    matrix = merge_permissions(role, permissions)
    return bool(matrix.get(module, {}).get(action))


def effective_role_for_owner() -> str:
    return "store_owner"
